### PR TITLE
Add PHP Composer support

### LINKED ISSUE #136: analyzer: Add PHP Composer support
Add support for the PHP Composer package manger to the analyer module.

https://getcomposer.org/

### PR DESCRIPTION
**Pull request by pnowakowski**
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/170)

### CODE REVIEW

**Comment by sschuberth** [comment]
Nit: Mind putting only "Add PHP Composer support" into the commit messager title, and put "Resolve**s** #136." into the commit message body?

**Comment by mnonnenmacher** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[analyzer/src/funTest/kotlin/PhpComposerTest.kt, line 54 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170#-L0toEIJ0dqNanqOccog:-L0toEIKXILYwUtS67yk:bmygprp) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/3d2012a8fc5677f2df33087396492ecb39e43e46/analyzer/src/funTest/kotlin/PhpComposerTest.kt#L54)):*
> ```Kotlin
>             result.project.scopes.size shouldBe 2
>             result.packages.size should beGreaterThan(0)
>             result.hasErrors() shouldBe false
> ```

I'd like to have this test aligned with the other PackageManager tests like NpmTest or GradleTest. It should have a yml fixture with the expected results.

---

*[analyzer/src/funTest/kotlin/PhpComposerTest.kt, line 89 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170#-L0toOb2HuU8mxv0r9Kl:-L0toOb2HuU8mxv0r9Km:bd7uqy6) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/3d2012a8fc5677f2df33087396492ecb39e43e46/analyzer/src/funTest/kotlin/PhpComposerTest.kt#L89)):*
> ```Kotlin
>             }
>             outputDir.deleteRecursively()
>         }.config(tags = setOf(Expensive))
> ```

This test should go to a separate test class DrupalIntegrationTest, see for example GradleIntegrationTest.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170)*

**Comment by pnowakowski** [comment]
Done.

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[analyzer/src/funTest/kotlin/PhpComposerTest.kt, line 54 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170#-L0toEIJ0dqNanqOccog:-L0xE3gmCEQjPTeVtY4R:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/3d2012a8fc5677f2df33087396492ecb39e43e46/analyzer/src/funTest/kotlin/PhpComposerTest.kt#L54)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

I'd like to have this test aligned with the other PackageManager tests like NpmTest or GradleTest. It should have a yml fixture with the expected results.
</blockquote></details>

Done.

---

*[analyzer/src/funTest/kotlin/PhpComposerTest.kt, line 89 at r1](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170#-L0toOb2HuU8mxv0r9Kl:-L0xg5NOa_uNs03IwS1y:b-896fix) ([raw file](https://github.com/heremaps/oss-review-toolkit/blob/3d2012a8fc5677f2df33087396492ecb39e43e46/analyzer/src/funTest/kotlin/PhpComposerTest.kt#L89)):*
<details><summary><i>Previously, mnonnenmacher (Martin Nonnenmacher) wrote…</i></summary><blockquote>

This test should go to a separate test class DrupalIntegrationTest, see for example GradleIntegrationTest.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/heremaps/oss-review-toolkit/170#-:-L0xE2KymRlyCJUW2wS5:b-896fix)*

**Comment by haikoschol** [comment]
@mnonnenmacher @sschuberth This PR is ready for review now. I changed the way information about packages is gathered (reading `composer.json`/`composer.lock` instead of calling `composer show <package>`) and added the usual checks for a lock file and a known-good Composer version.

Two issues I'd like to bring up for the review are:

1. Package metadata has an attribute `dist` with a URL to an archive. Consistent with other package manager implementation, these are used as the "binary artifact" for the package. The `dist` object has a field called `shasum`, but many (maybe all) packages do not provide a value for it. Do we still want to store the URL as binary artifact?

2. Package names have the format `clue/stream-filter`. Currently the whole string is stored as `name` for the package, but we could also use the first part as `namespace`.

**Comment by sschuberth** [comment]
> The `dist` object has a field called `shasum`, but many (maybe all) packages do not provide a value for it. Do we still want to store the URL as binary artifact?

I'd say yes. It's more important for us to actually get an artifact, than (also) being able to verify its integrity.

> Package names have the format `clue/stream-filter`. Currently the whole string is stored as `name` for the package, but we could also use the first part as `namespace`.

If the part before the slash *is* actually used to group packages, e.g. by vendor or topic, and I'd say yes, we should use it as a namespace. What is the semantic meaning of the part before the slash (if any)?

**Comment by mnonnenmacher** [comment]
> I'd say yes. It's more important for us to actually get an artifact, than (also) being able to verify its integrity.

I agree, for source artifacts the checksum verification is mandatory, but for binary artifacts it's better to have the URL than nothing. We could download the artifact to calculate the checksum, but that's likely too expensive for the analyzer.

**Comment by haikoschol** [comment]
> If the part before the slash is actually used to group packages, e.g. by vendor or topic, and I'd say yes, we should use it as a namespace. What is the semantic meaning of the part before the slash (if any)?

The documentation calls the part before the slash "[vendor name](https://getcomposer.org/doc/04-schema.md#name)". I'll change the code to use it as the namespace.

File: .travis.yml [hunk 1]
```diff
@@ -28,6 +28,8 @@ install:
   - sudo apt install -y cvs
   - eval "$(gimme)"
   - curl https://raw.githubusercontent.com/golang/dep/7d5cd199ce454707f81c63b7ea4299151b8b981d/install.sh | sh
+  - gem install licensee -v 9.8.0
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply .travis.yml]: Done.

File: .travis.yml [hunk 2]
```diff
@@ -28,6 +28,8 @@ install:
...
+  - phpenv global 7.1
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply .travis.yml]: Still an open question for @haikoschol.

File: analyzer/src/funTest/kotlin/PhpComposerTest.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2017 HERE Europe B.V.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/PhpComposerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/PhpComposerTest.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package com.here.ort.analyzer
+
+import com.here.ort.analyzer.managers.PhpComposer
+import com.here.ort.downloader.VersionControlSystem
+import com.here.ort.model.yamlMapper
+import com.here.ort.utils.normalizeVcsUrl
+
+import io.kotlintest.matchers.shouldBe
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/PhpComposerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/PhpComposerTest.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import io.kotlintest.specs.StringSpec
+
+import java.io.File
+
+private val projectDir = File("src/funTest/assets/projects/synthetic/php-composer")
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/PhpComposerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/integration/DrupalIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import com.here.ort.analyzer.managers.PhpComposer
+import com.here.ort.model.Identifier
+import com.here.ort.model.Package
+import com.here.ort.model.RemoteArtifact
+import com.here.ort.model.VcsInfo
+
+import java.io.File
+
+class DrupalIntegrationTest : AbstractIntegrationSpec() {
+
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/integration/DrupalIntegrationTest.kt]: Done.

File: analyzer/src/funTest/kotlin/integration/DrupalIntegrationTest.kt [hunk 7]
```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2017 HERE Europe B.V.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/integration/DrupalIntegrationTest.kt]: Done.

File: analyzer/src/funTest/kotlin/PhpComposerTest.kt [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            val result = PhpComposer.create().resolveDependencies(listOf(packageFile))[packageFile]
+            val expectedResults = patchExpectedResult(File(projectDir.parentFile, "php-composer-expected-output.yml")
+                    .readText())
+            yamlMapper.writeValueAsString(result) shouldBe expectedResults
+        }
+    }
+
+    private fun patchExpectedResult(result: String) =
+            //vcs:
+            result.replaceFirst("url: \"\"", "url: \"$vcsUrl\"")
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/PhpComposerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/PhpComposerTest.kt [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+private val projectDir = File("src/funTest/assets/projects/synthetic/php-composer")
+
+class PhpComposerTest : StringSpec() {
+    private val vcsDir = VersionControlSystem.forDirectory(projectDir)!!
+    private val vcsRevision = vcsDir.getRevision()
+    private val vcsUrl = vcsDir.getRemoteUrl()
+
+    init {
+        "Php composer recognises project" {
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/funTest/kotlin/PhpComposerTest.kt]: Done.

File: .travis.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (haikoschol) [reply .travis.yml]: @sschuberth No idea. I left `.travis.yml` and `.appveyor.yml` untouched.

**Pull request merged**